File: core/src/com/mygdx/game/Bodies/MoveableImage.java
package com.mygdx.game.Bodies;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.mygdx.game.Extra.AssemblingScreenCoords;
import com.mygdx.game.Extra.ItemID;

public class MoveableImage extends Image implements ItemID, AssemblingScreenCoords {

    private static final float PIdiv2 = (float) Math.PI / 2;

    public boolean isMoving = false;
    public boolean isTouchable = false;
    private float x;
    private float y;
    private float width;
    private float height;
    private float angle;
    private float startPosX;
    private float startPosY;
    private int XinTable;
    private int YinTable;

    private int itemID;
    private int itemFacing;


    public MoveableImage(float pos_x, float pos_y, float aWidth, float aHeight, float angle, int ID) {
        super(new Texture(getImageName(ID)));
        XinTable = 50;
        YinTable = 50;
        startPosX= pos_x;
        startPosY = pos_y;
        this.x = pos_x;
        this.y = pos_y;
        this.width = aWidth;
        this.height = aHeight;
        this.angle = angle;
        this.itemID = ID;
        switch (ID) {
            case HALF_WOOD_BLOCK: this.itemFacing = FACING_DOWNLEFT; break;
            case HALF_STEEL_BLOCK: this.itemFacing = FACING_DOWNLEFT; break;
            default: this.itemFacing = MISSING;
        }
    }

    public int getXinTable() {
        return XinTable;
    }
    public void setXinTable(int xinTable) {
        XinTable = xinTable;
    }

    public int getYinTable() {
        return YinTable;
    }
    public void setYinTable(int yinTable) {
        YinTable = yinTable;
    }

    @Override
    public float getX() {
        return x;
    }
    @Override
    public void setX(float x) {
        this.x = x;
    }

    @Override
    public float getY() {
        return y;
    }
    @Override
    public void setY(float y) {
        this.y = y;
    }

    @Override
    public float getWidth() {
        return width;
    }
    @Override
    public void setWidth(float width) {
        this.width = width;
    }

    @Override
    public float getHeight() {
        return height;
    }
    @Override
    public void setHeight(float height) {
        this.height = height;
    }

    private float getAngle() {
        return angle;
    }
    private void setAngle(float angle) {
        this.angle = angle;
    }

    public int getItemFacing() {
        return itemFacing;
    }
    public void setItemFacing(int itemFacing) {
        this.itemFacing = itemFacing;
    }


    @Override
    public void draw(Batch batch, float parentAlpha) {
        super.draw(batch, parentAlpha);
    }
    @Override
    public void act(float delta) {
        super.act(delta);
        this.setRotation(angle*  MathUtils.radiansToDegrees);
        this.setPosition(x,y);
    }


    public boolean contains(float x, float y){

        return (x>this.x && x<(this.x+this.width)) && (y< Gdx.graphics.getHeight()-this.y && y>Gdx.graphics.getHeight()-this.y-this.height);
    }

    public void returnToStartPos(){
        this.setX(startPosX);
        this.setY(startPosY);
    }

    public boolean isInStartPos(){
        return x == startPosX && y == startPosY;
    }



    private static String getImageName(int ID) {
        switch(ID){
            case EMPTY_CELL: return "gray.png";
            case WOOD_BLOCK: return "woodblock.png";
            case STEEL_BLOCK: return "steelblock.png";
            case ENGINE: return "engine.png";
            case TURBINE: return "turbine.png";
            case HALF_WOOD_BLOCK: return "halfwoodblock.png";
            case HALF_STEEL_BLOCK: return "halfsteelblock.png";
            case GUN_1: return "gun_1.png";
            case GUN_2: return "gun_2.png";
        }
        return "";
    }

    public void rotate() {
        if (this.itemID == HALF_WOOD_BLOCK || this.itemID == HALF_STEEL_BLOCK) {
            this.setAngle(this.getAngle() - PIdiv2);
            System.out.println("-");
            switch (this.getItemFacing()) {
                case FACING_DOWNLEFT: {
                    this.setItemFacing(FACING_UPLEFT);
                    this.setY(this.getY() + BLOCK_SIZE);
                    System.out.println("-1");
                    break;
                }
                case FACING_UPLEFT: {
                    this.setItemFacing(FACING_UPRIGHT);
                    this.setX(this.getX() + BLOCK_SIZE);
                    System.out.println("-2");
                    break;
                }
                case FACING_UPRIGHT: {
                    this.setItemFacing(FACING_DOWNRIGHT);
                    this.setY(this.getY() - BLOCK_SIZE);
                    System.out.println("-3");
                    break;
                }
                case FACING_DOWNRIGHT: {
                    this.setItemFacing(FACING_DOWNLEFT);
                    this.setX(this.getX() - BLOCK_SIZE);
                    System.out.println("-4");
                    break;
                }
            }
        }
    }

    public void returnBaseRotation() {
        if (this.itemID == HALF_WOOD_BLOCK || this.itemID == HALF_STEEL_BLOCK) {
            System.out.println("+");
            switch (this.getItemFacing()) {
                case FACING_UPLEFT: {
                    this.setAngle(this.getAngle() + PIdiv2);
                    System.out.println("+1");
                    break;
                }
                case FACING_UPRIGHT: {
                    this.setAngle(this.getAngle() + 2*PIdiv2);
                    System.out.println("+2");
                    break;
                }
                case FACING_DOWNRIGHT: {
                    this.setAngle(this.getAngle() - PIdiv2);
                    System.out.println("+3");
                    break;
                }
            }
            this.setItemFacing(FACING_DOWNLEFT);
        }
    }

}

File: core/src/com/mygdx/game/Bodies/MoveableImage.kt
package com.mygdx.game.Bodies

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.mygdx.game.Extra.ItemID.*
import com.mygdx.game.Extra.AssemblingScreenCoords.*


class MoveableImage(private val startPosX: Float, private val startPosY: Float, private var width: Float, private var height: Float, var angle: Int, texture: String) : Image(Texture(texture)) {

    var isMoving = false
    private var isTouchable = false
    var isAlreadyMoved = false
    private var x: Float = 0.toFloat()
    private var y: Float = 0.toFloat()
    private val number: Int
    var xinTable: Int = 0
    var yinTable: Int = 0


    val isInStartPos: Boolean
        get() = x == startPosX && y == startPosY

    init {
        number = getString(texture)
        xinTable = NULL
        yinTable = NULL
        this.x = startPosX
        this.y = startPosY
        if (number != TURBINE)
            this.setOrigin(BLOCK_SIZE / 2, height / 2)
        else
            this.setOrigin(width - BLOCK_SIZE / 2, height / 2)
    }

    override fun isTouchable(): Boolean {
        return isTouchable
    }
    fun setTouchable(touchable: Boolean) {
        isTouchable = touchable
    }

    private fun getString(texture: String): Int {
        return when (texture) {
            "woodblock.png" -> WOOD_BLOCK
            "steelblock.png" -> STEEL_BLOCK
            "engine.png" -> ENGINE
            "turbine.png" -> TURBINE
            "halfwoodblock.png" -> HALF_WOOD_BLOCK
            "halfsteelblock.png" -> HALF_STEEL_BLOCK
            "gun_1.png" -> WOOD_GUN
            "gun_2.png" -> STEEL_GUN
            else -> NULL
        }
    }

    fun getNumber(): Int {
        when (angle) {
            0 -> return number
            90 -> return number + UP
            180 -> return number + LEFT
            270 -> return number + DOWN
        }
        return number
    }

    override fun getX(): Float {
        return x
    }
    override fun setX(x: Float) {
        this.x = x
    }

    override fun getY(): Float {
        return y
    }
    override fun setY(y: Float) {
        this.y = y
    }

    fun setXY(x: Float, y: Float) {
        this.setX(x)
        this.setY(y)
    }
    fun setXYinTable(x: Int, y: Int) {
        this.xinTable = x
        this.yinTable = y
    }

    override fun getWidth(): Float {
        return width
    }
    override fun setWidth(width: Float) {
        this.width = width
    }

    override fun getHeight(): Float {
        return height
    }
    override fun setHeight(height: Float) {
        this.height = height
    }

    override fun act(delta: Float) {
        super.act(delta)
        this.rotation = angle.toFloat()
        this.setPosition(x, y)
    }

    fun contains(x: Float, y: Float): Boolean {
        when (angle) {
            0 -> return (x > this.x) &&
                    (x < this.x + this.width) &&
                    (y < SCREEN_HEIGHT - this.y) &&
                    (y > SCREEN_HEIGHT_F - this.y - this.height)
            90 -> return (x > this.x + originX - this.height + originY) &&
                    (x < this.x + originX + originY) &&
                    (y < SCREEN_HEIGHT_F - this.y - originY + originX) &&
                    (y > SCREEN_HEIGHT_F - this.y - originY + originX - this.width)
            180 -> return (x > this.x + 2 * originX - this.width) &&
                    (x < this.x + 2 * originX) &&
                    (y < SCREEN_HEIGHT - this.y) &&
                    (y > SCREEN_HEIGHT_F - this.y - this.height)
            270 -> return (x > this.x + originX - originY) &&
                    (x < this.x + originX + this.height - originY) &&
                    (y < SCREEN_HEIGHT_F - this.y - originY - originX + this.width) &&
                    (y > SCREEN_HEIGHT_F - this.y - originY - originX)
        }
        return false
    }

    fun returnToStartPos(currentX: Float, currentY: Float) {
        this.setX(currentX)
        this.setY(currentY)
    }
    fun returnToStartPos() {
        this.setX(startPosX)
        this.setY(startPosY)
    }

    fun flip90() {
        angle += 90
        if (angle == 360) angle = 0
    }


}
